checkmarx/checkmarx.py

```diff
@@ -0,0 +1,293 @@
+
+    return model.FinishedScans(
+        scans_above_threshold=scan_results_above_threshold,
+        scans_below_threshold=scan_results_below_threshold,
+        failed_components=failed_components,
+    )
+
+
+def send_mail(
+    scans: model.FinishedScans,
+    threshold: int,
+    email_recipients,
+    routes: checkmarx.client.CheckmarxRoutes,
+):
+    body = checkmarx.tablefmt.assemble_mail_body(
+        scans_above_threshold=scans.scans_above_threshold,
+        scans_below_threshold=scans.scans_below_threshold,
+        failed_components=scans.failed_components,
+        threshold=threshold,
+        routes=routes,
+    )
+    try:
+        # get standard cfg set for email cfg
+        default_cfg_set_name = ci.util.current_config_set_name()
+        cfg_factory = ci.util.ctx().cfg_factory()
+        cfg_set = cfg_factory.cfg_set(default_cfg_set_name)
+
+        ci.util.info(f'sending notification emails to: {",".join(email_recipients)}')
+        mailutil._send_mail(
+            email_cfg=cfg_set.email(),
+            recipients=email_recipients,
+            mail_template=body,
+            subject='[Action Required] checkmarx vulnerability report',
+            mimetype='html',
+        )
+        ci.util.info('sent notification emails to: ' + ','.join(email_recipients))
+
+    except Exception:
+        traceback.print_exc()
+        ci.util.warning('error whilst trying to send notification-mail')
+
+
+def print_scans(
+    scans: model.FinishedScans,
+    routes: checkmarx.client.CheckmarxRoutes,
+):
+    # XXX raise if an error occurred?
+    if scans.scans_above_threshold:
+        print('\n')
+        ci.util.info('Critical scans above threshold')
+        checkmarx.tablefmt.print_scan_result(
+            scan_results=scans.scans_above_threshold,
+            routes=routes,
+        )
+    else:
+        ci.util.info('no critical components above threshold found')
+
+    if scans.scans_below_threshold:
+        print('\n')
+        ci.util.info('Clean scans below threshold')
+        checkmarx.tablefmt.print_scan_result(
+            scan_results=scans.scans_below_threshold,
+            routes=routes,
+        )
+    else:
+        ci.util.info('no scans below threshold to print')
+
+    if scans.failed_components:
+        print('\n')
+        failed_components_str = '\n'.join(
+            (
+                component_name for component_name in scans.failed_components
+            )
+        )
+        ci.util.info(f'failed components:\n{failed_components_str}')
+
+
+def _download_and_zip_repo(repo, ref: str,path_filter_func: typing.Callable, tmp):
+    url = repo._build_url('tarball', ref, base_url=repo._api)
+    with repo._get(url, stream=True) as r, \
+        tarfile.open(fileobj=r.raw, mode='r|*') as tar, \
+        zipfile.ZipFile(tmp, mode='w', compression=zipfile.ZIP_DEFLATED) as zip_file:
+
+        r.raise_for_status()
+        r.raw.seekable = False
```

Comment:
nit: is is really necessary to patch-in the `seekable` attr, when passing `|` to the tarfile c'tor?

Reply:
tested both and both seemed to work fine. Since it's already a stream python should know that it's not seekable